cloud_functions/test/flagCryptominer-error-tests.js

```diff
@@ -0,0 +1,88 @@
+// These tests cover various error cases for the flagCryptominer cloud function.
+// Based on https://cloud.google.com/functions/docs/testing/test-background#unit_tests
+//
+// Since this is a background cloud function, errors are exposed via Cloud Logging Error Reporting.
+// Uncaught exceptions will log errors, but will also incur a cold start on future invocations.
+// Instead, we use console.error, so we need to stub that function for these tests.
+// See https://cloud.google.com/functions/docs/monitoring/error-reporting for details.
+//
+// IMPORTANT: Since we're stubbing a global function (console.error), all of these tests must run
+// in serial, not parallel as is Ava's default.
+// (For future exploration: https://github.com/sinonjs/sinon-test)
+const test = require('ava')
+const flagCryptominer = require('../flagCryptominer')
+const sinon = require('sinon')
+const utils = require('./test-utils')
+
+
+test.beforeEach(t => {
+  sinon.stub(console, 'error')
+})
+
+test.afterEach.always(t => {
+  console.error.restore()
+})
+
+
+// The function should do nothing if the cryptominer attribute isn't set. This could indicate a
+// misconfigured Pub/Sub topic subscription.
+test.serial('reject non-cryptominer messages', async t => {
+  const mixpanel = utils.makeStubMixpanel()
+  const message = {
+    data: utils.encodeMessageData({ userSubjectId: '123' })
+  }
+
+  await flagCryptominer.run(mixpanel, message)
+
+  checkConsoleError(t, /unexpected message/i)
+  t.assert(mixpanel.people.set.notCalled)
+})
+
+test.serial('data is not JSON', async t => {
```

Comment:
What's an example of a potential future bug that might be introduced without this test, and what would be the negative impact of that bug?

Reply:
A message publisher sending a malformed message that it thinks is correct. Due to the async nature, the message publisher does not get direct feedback from this function, so I want to make sure we detect that and not be silent about it. The impact if this is silent is that we think we're capturing data but we're not, and we would only notice this after looking at reports in Mixpanel, if we even notice at all.